README.md
# XDApp RPC Service SDK

XDAppRPC服务SDK，使用方法

```js
const xdapp = require('xdapp');

// 其中 demo 为项目名，test 为服务名，123456 为密钥
const service = new xdapp('demo', 'test', '123456');

// 注册一个方法
service.register(function(args) {
    console.log(args);
    return true;
}, 'abc');

// 等同于上面的方法, 其中 test_ 为服务前缀
// 请注意，只有服务名相同的前缀rpc方法才会被外部调用到
service.addFunction(function(args) {
    // 获取 context 对象
    const context = xdapp.getCurrentContext();
    console.log(context.adminId);

    console.log(args);
    return true;
}, 'test_abc');

// 连接到本地测试开发服务器
service.connectToLocalDev('127.0.0.1', 8082);

// 连接到外网测试服务器
// service.connectToDev();

// 连接到生产环境(国内项目)
// service.connectToProduce();

// 连接到生产环境(海外项目)
// service.connectToGlobal();
```

更多的使用方法see: [https://github.com/hprose/hprose-nodejs/wiki/Hprose-服务器](https://github.com/hprose/hprose-nodejs/wiki/Hprose-服务器)



`ServiceAgent` 接受3个参数，分别是应用英文名、服务名和密钥，应用是在 https://www.xdapp.com/ 里创建的应用，服务是在应用后台内自行创建的服务名，密钥是对应每个应用的连接密钥

### 关于 `context` 上下文对象

在RPC请求时，如果需要获取到请求时的管理员ID等等参数，可以用此获取，如上面 `hello` 的例子，通过 `context = xdapp.getCurrentContext()` 可获取到 `context`，包括：

参数         |   说明
------------|---------------------
service     | 当前服务
client      | 通信的连接对象，可以使用 `close()` 方法关闭连接
requestId   | 请求的ID
appId       | 请求的应用ID
serviceId   | 请求发起的服务ID，0表示XDApp系统请求，1表示来自浏览器的请求
adminId     | 请求的管理员ID，0表示系统请求
userdata    | 默认 {} 对象，可以自行设置参数

返回的 `service` 对象常用方法如下：

### `connectToProduce()`

连接到国内生产环境，将会创建一个异步tls连接接受和发送RPC数据，无需自行暴露端口，如果遇到网络问题和服务器断开可以自动重连，除非是因为密钥等问题导致的断开将不会重新连接

### `connectToGlobal()`

连接到海外生产环境，同 `connectToProduce()` 区别在于项目是海外的

### `connectToDev(serviceKey = None)`

同上，连接到研发环境, 不设置 serviceKey 则使用 new ServiceAgent 时传入的密钥

### `connectToLocalDev(self, host = '127.0.0.1', port = 8061, serviceKey = null)`

同上，连接到本地研发服务器，请下载 XDApp-Console-UI 服务包，https://hub000.xindong.com/core-system/xdapp-console-ui ，启动服务



### `addHttpApiProxy(url, alias = 'api', methods = ['get'], array httpHeaders = [])`

添加一个http代理

使用场景：
当服务器里提供一个内部Http接口，但是它没有暴露给外网也没有权限验证处理，但希望Web页面可以使用
此时可以使用此方法，将它暴露成为一个XDApp的RPC服务，在网页里直接通过RPC请求将数据转发到SDK请求后返回，不仅可以实现内网穿透功能还可以在Console后台设置访问权限。

每个Http代理请求都会带以下头信息，方便业务处理:

* X-Xdapp-Proxy: True
* X-Xdapp-App-Id: 1
* X-Xdapp-Service: demo
* X-Xdapp-Request-Id: 1
* X-Xdapp-Admin-Id: 1

```javascript
service.addHttpApiProxy('http://127.0.0.1:9999', 'myApi', ['get', 'post', 'delete', 'put'])
```

Vue页面使用

方法接受3个参数，uri, data, timeout，其中 data 只有在 post 和 put 有效，timeout 默认 30 秒

```javascript
// 其中gm为注册的服务名
this.$service.gm.myApi.get('/uri?a=arg1&b=arg2');
// 最终将会请求 http://127.0.0.1:9999/uri?a=arg1&b=arg2
// 返回对象 {code: 200, headers: {...}, body: '...'}

// 使用post, 第2个参数接受string或字符串, 第3个参数可设置超时
this.$service.gm.myApi.post('/uri?a=1', {a:'arg1', b:'arg2'}, 15);
```

### 同时连接多个环境

一个 `service` 可以同时连接 `connectToProduce`, `connectToDev`, `connectToLocalDev` 3个，但需要保证使用正确的密钥。但不建议将测试环境的连接到生产环境服务器里

### `register(function, alias = null, resultMode = HproseResultMode.Normal, simple = null)`

注册一个RPC方法到服务上，它是 `service.addFunction()` 方法的封装，差别在于会自动对 `alias` 增加 `serviceName` 前缀

`register.register(hello, 'hello')` 相当于 `register.addFunction(hello, 'servicename_hello')`

### `addFunction(function, alias = null, resultMode = HproseResultMode.Normal, simple = null)`

注册一个RPC方法到服务上


### `addMissingFunction(function, resultMode = HproseResultMode.Normal, simple = null)`

此方法注册后，所有未知RPC请求都降调用它，它将传入2个参数，分别是RPC调用名称和参数

### `addFilter() / removeFilter()` 过滤器

可以方便开发调试

see [https://github.com/hprose/hprose-nodejs/wiki/Hprose-过滤器](https://github.com/hprose/hprose-nodejs/wiki/Hprose-过滤器)

## 常见问题

* 使用 addInstanceMethods() 或hprose的方法暴露rpc方法，无法调用方法，提示方法不存在<br>
  不是使用 register() 方法暴露rpc方法，则必须加上服务前缀，例如，应该用 `service.addInstanceMethods(obj, serviceName)`

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.2.0](https://github.com/xdapp/xdapp-sdk-nodejs/compare/v1.0.12...v1.2.0) (2020-03-13)


### Features

* **release:** add standard-version ([8b1b807](https://github.com/xdapp/xdapp-sdk-nodejs/commit/8b1b807b7132f4bd5c9cc16344dfe10b41147c9b))
* **ts:** 改为 typescript 新增 dts 类型提示 ([aa28eb4](https://github.com/xdapp/xdapp-sdk-nodejs/commit/aa28eb47da03fcaa196630d5b7fc171ddacea01e))


### Bug Fixes

* **deps:** 调整依赖 ([46decc1](https://github.com/xdapp/xdapp-sdk-nodejs/commit/46decc175a93626b7f55035e5a16e07a00df6289))


src/xdapp.ts
import tls from "tls";
import net from "net";
import fetch from "node-fetch";
import hprose from "hprose";
import FormData from "form-data";
import { parse } from "url";
import debug from "debug";
import SysService from "./sysService";

const debugXD = debug("xdapp:sdk");

const hproseService = hprose.Service;
const Future = hprose.Future;

const ProductionServer = {
  host: "service-prod.xdapp.com",
  port: 8900
};
const DevServer = {
  host: "dev.xdapp.com",
  port: 8100
};
const GlobalServer = {
  host: "service-gcp.xdapp.com",
  port: 8900
};

const RPC_VERSION = 1; // RPC协议版本
const FLAG_RESULT_MODE = 2; // 返回模式
const FLAG_FINISH = 4; // 已完成

type ConnectOption = {
  tls: boolean;
  localDev: boolean;
  dev: boolean;
};

class XDAppServiceAgent extends hproseService {
  private showLog: boolean;
  private serviceName: string;
  private appName: string;
  constructor(appName: string, serviceName: string, serviceKey = "") {
    if (!appName) {
      throw new Error("Required appName");
    }
    if (!serviceName) {
      throw new Error("Required serviceName");
    }
    super();
    this.showLog = false;
    this.serviceName = serviceName;
    this.appName = appName;

    // 注册系统服务
    this.addInstanceMethods(
      SysService(this, appName, serviceName, serviceKey),
      "sys"
    );
  }

  addHttpApiProxy(
    url: string,
    alias = "api",
    methods = ["get"]
  ) {
    const self = this;
    if (typeof methods === "string") methods = [methods];
    methods.forEach(function(method) {
      self.addWebFunction(
        function(uri: string, data?: any, timeout: number = 30, callback?: (response: any) => void) {
          const context = self.getCurrentContext();

          let urlParse = parse(url + uri);
          const options = {
            uri: urlParse,
            method: method.toUpperCase(),
            timeout: timeout * 1000,
            encoding: "utf8",
            gzip: true,
            headers: {
              // 'Host': urlParse.hostname,
              "User-Agent": "Chrome/49.0.2587.3",
              "X-Xdapp-Proxy": "True",
              "X-Xdapp-App-Id": context.appId,
              "X-Xdapp-Service": self.serviceName,
              "X-Xdapp-Request-Id": context.requestId,
              "X-Xdapp-Admin-Id": context.adminId,
              "Content-Type": "application/x-www-form-urlencoded"
            },
            body: ""
          };

          if (options.method === "POST") {
            options.body =
              typeof data === "object" ? new FormData(data) : data.toString();
          } else if (options.method === "PUT") {
            options.body =
              typeof data === "object" ? JSON.stringify(data) : "" + data;
          }

          // TODO: 不确定是否一定为 json
          fetch(urlParse, options)
            .then(x => x.json())
            .then(resp => {
              const response = {
                code: resp.status,
                headers: resp.headers || {},
                body: resp.body
              };

              if (!resp) {
                response.code = -1;
              }

              debugXD(
                "Http Proxy, method: " +
                  method +
                  ", code: " +
                  resp.statusCode +
                  ", url: " +
                  url +
                  uri +
                  ", headers: " +
                  JSON.stringify(resp.headers || {})
              );
              return callback && callback(response);
            });
        },
        alias + "_" + method,
        { async: true }
      );
    });
  }

  // 注册一个Web浏览器可用的方法，同 hprose.addFunction 方法参数，差别是默认增加了 alias 了 serviceName 的前缀
  addWebFunction(func: (uri: string, data?: any, timeout?: number, callback?: () => void) => void, alias: string, options: any) {
    alias = this.serviceName + "_" + alias;
    this.addFunction(func, alias, options);
  }

  // deprecated 请使用 addWebFunction
  register(func: () => void, alias: string, options: any) {
    this.addWebFunction(func, alias, options);
  }

  // 获取当前上下文对象
  getCurrentContext() {
    return this._lastContext;
  }

  // 连接到本地测试环境
  connectToLocalDev(host = "127.0.0.1", port = 8061) {
    return this.connectTo(host, port, {
      tls: false,
      localDev: true,
      dev: true
    });
  }

  // 连接到测试环境
  connectToDev() {
    return this.connectTo(null, null, {
      tls: true,
      localDev: false,
      dev: true
    });
  }

  // 连接到国内生产环境
  connectToProduce() {
    return this.connectTo(null, null, {
      tls: true,
      localDev: false,
      dev: false
    });
  }
  // 连接到海外生产环境
  connectToGlobal() {
    return this.connectTo(GlobalServer.host, GlobalServer.port, {
      tls: true,
      localDev: false,
      dev: false
    });
  }

  // 创建一个新的连接
  connectTo(host: string | null, port: number | null, option: ConnectOption) {
    if (this.socket) return false;

    option = Object.assign(
      {
        tls: true,
        localDev: false,
        dev: false,
        serviceKey: null
      },
      option || {}
    );

    host = host || (option.dev ? DevServer.host : ProductionServer.host);
    port = port || (option.dev ? DevServer.port : ProductionServer.port);

    const socket = (option.tls ? tls : (net as any)).connect(
      port,
      host,
      {},
      (err: any) => {
          debugXD(port, host, err)
        debugXD(
          "Rpc client connected. " +
            (socket.authorized ? "authorized" : "unauthorized")
        );
      }
    );
    socket.on("data", (data: Buffer) => {
      // CFlag/CVer/NLength/NAppId/NServiceId/NRequestId/NAdminId/CContextLength
      if (data.length < 20) {
        debugXD(`接收到的数据包异常`);
        socket.end();
        return;
      }
      const flag = data.readInt8(0);
      const ver = data.readInt8(1);
      if (ver !== RPC_VERSION) {
        debugXD(`当前协议版本不被支持 ${ver}`);
        socket.end();
        return;
      }

      const length = data.readInt32BE(2);
      const contextLength = data.readInt8(22);
      const request = {
        flag: flag,
        ver: ver,
        length: length,
        appId: data.readInt32BE(6),
        serviceId: data.readInt32BE(10),
        requestId: data.readInt32BE(14),
        adminId: data.readInt32BE(18),
        contextLength: contextLength,
        context: contextLength > 0 ? data.slice(23, 23 + contextLength) : "",
        body: data.slice(contextLength + 23)
      };

      const context = {
        service: this,
        client: socket,
        requestId: request.requestId,
        appId: request.appId,
        serviceId: request.serviceId,
        adminId: request.adminId,
        userdata: {}
      };
      this._lastContext = context;
      const rs = this.defaultHandle(request.body.toString(), context);

      if (Future.isFuture(rs)) {
        rs.then(function(data: Buffer) {
          _send(socket, data, request);
        });
      } else {
        _send(socket, data, request);
      }
    });
    socket.on("error", (error: Error) => {
      debugXD("socket error: ", error);
      this.socket = null;
      this.regSuccess = false;

      if (!this.isRegError) {
        setTimeout(() => {
          // isRegError
          this.connectTo(host, port, option);
        }, 2000);
      }
    });
    socket.on("close", () => {
      debugXD(
        "Rpc server close connection" +
          (this.isRegError ? "" : ", reconnect after 1 second.")
      );
      this.socket = null;
      this.regSuccess = false;

      if (!this.isRegError) {
        setTimeout(() => {
          // isRegError
          this.connectTo(host, port, option);
        }, 1000);
      }
    });

    // // 超时重置
    // socket.setTimeout(30 * 1000);
    // socket.on('timeout', () => {
    //     this.log('Rpc socket timeout.');
    //     socket.end();
    // });

    function _send(socket: any, body: any, request: any) {
      // CFlag/CVer/NLength/NAppId/NServiceId/NRequestId/NAdminId/CContextLength
      const packagePrefix = 6;
      const headerLength = 17;
      const length = headerLength + request.contextLength + body.length;

      const bufHeader = Buffer.alloc(packagePrefix + headerLength);
      // packagePrefix
      const flag = request.flag | FLAG_RESULT_MODE | FLAG_FINISH;
      bufHeader.writeInt8(flag, 0); // flag
      bufHeader.writeInt8(RPC_VERSION, 1); // ver
      bufHeader.writeInt32BE(length, 2); // length
      // header
      bufHeader.writeInt32BE(request.appId, 6); // appId
      bufHeader.writeInt32BE(request.serviceId, 10); // serviceId
      bufHeader.writeInt32BE(request.requestId, 14); // requestId
      bufHeader.writeInt32BE(request.adminId, 18); // adminId
      bufHeader.writeInt8(request.contextLength, 22); // contextLength

      const buffers = [bufHeader];
      if (request.contextLength > 0) {
        // 将自定义内容加入
        buffers.push(request.context);
      }
      buffers.push(body);
      const buf = Buffer.concat(
        buffers,
        bufHeader.length + request.contextLength + body.length
      );
      socket.write(buf);
    }

    this.socket = socket;
    return socket;
  }

  public log(msg: string) {
      debugXD(msg)
  }

  // TODO: 不知道是啥，可能被父方法调用？
  setLogHandle(handle: any) {
    this.log = handle;
  }
}

export default XDAppServiceAgent


src/sysService.js
const crypto = require('crypto');

function sha1(str) {
    let hmac = crypto.createHash('sha1');
    hmac.update(str);
    return hmac.digest('hex');
}

// 设置通信配置
const sysService = function (XDAppServiceAgent, appName, serviceName, serviceKey) {
    const sysCall = {};

    function getHash(time, rand) {
        return sha1(`${appName}.${serviceName}.${time}.${rand}.${serviceKey}.xdapp.com`);
    }

    sysCall.reg = function (time, rand, hash) {
        if (XDAppServiceAgent.regSuccess) return {
            'status': false
        }
        XDAppServiceAgent.isRegError = false;
        if (hash !== sha1(`${time}.${rand}.xdapp.com`)) {
            // 验证失败
            return {
                'status': false
            }
        }

        const now = parseInt(new Date().getTime() / 1000);
        if (Math.abs(now - time) > 60) {
            // 超时
            return {
                'status': false
            }
        }

        return {
            'status': true,
            'app': appName,
            'name': serviceName,
            'time': now,
            'rand': rand,
            'version': 'v1',
            'hash': getHash(now, rand),
        };
    };

    // 注册失败
    sysCall.regErr = function (msg, data = null) {
        XDAppServiceAgent.log(msg);
        XDAppServiceAgent.isRegError = true;
    }

    // 注册成功
    sysCall.regOk = function(data, time, rand, hash) {
        if (XDAppServiceAgent.regSuccess) return;
        const now = new Date().getTime() / 1000;
        if (Math.abs(now - time) > 60) {
            // 超时
            XDAppServiceAgent.log(`RPC验证超时，服务名: ${appName}->${serviceName}`);
            return false;
        }

        // 不接受低于16位长度的rand
        if (rand.length < 16) {
            XDAppServiceAgent.log(`regOk() 回调 rand 参数太短，服务名: ${appName}->${serviceName}`);
            return;
        }

        if (getHash(time, rand) !== hash) {
            // 验证失败
            XDAppServiceAgent.log(`RPC验证失败，服务名: ${appName}->${serviceName}`);
            XDAppServiceAgent.socket.close();
            return;
        }

        // 注册成功
        serviceData = data;
        XDAppServiceAgent.regSuccess = true;
        XDAppServiceAgent.serviceId = data.serviceId;

        XDAppServiceAgent.log(`RPC服务注册成功，服务名: ${appName}->${serviceName}`);

        let allService = {
            sys: [],
            service: [],
            other: [],
        };
        XDAppServiceAgent.getNames().forEach((item) => {
            if (item === '#')return null;
            let pos = item.indexOf('_');
            if (pos === -1) {
                allService.other.push(item);
                return null;
            }
            let type = item.substr(0, pos);
            let func = item.substr(pos + 1);
            switch (type) {
                case 'sys':
                    allService.sys.push(func.replace(/_/g, '.') +'()');
                    break;
                case serviceName:
                    allService.service.push(func.replace(/_/g, '.') +'()');
                    break;
                default:
                    allService.other.push(func.replace(/_/g, '.') +'()');
                    break;
            }
        });
        XDAppServiceAgent.log(`系统RPC：${allService.sys.join(', ')}`);
        XDAppServiceAgent.log(`已暴露服务RPC：${allService.service.join(', ')}`);
        if (allService.other.length > 0) {
            XDAppServiceAgent.log(`已暴露但XDApp不会调用的RPC：${allService.other.join(', ')}`);
            XDAppServiceAgent.log(`若需要这些方法暴露给XDApp服务使用，请加: ${serviceName} 前缀`);
        }
    };

    // rpc回调log输出
    sysCall.log = function(log, type, data = null) {
        XDAppServiceAgent.log(log, type, data);
    }

    // ping
    sysCall.ping = function() {
        return true;
    }

    // 获取服务器列表
    sysCall.getFunctions = function() {
        if (!XDAppServiceAgent.regSuccess)return [];
        return XDAppServiceAgent.getNames();
    }

    return sysCall;
}

module.exports = sysService;

src/sysService.ts
import crypto from "crypto";

function sha1(str: string) {
  let hmac = crypto.createHash("sha1");
  hmac.update(str);
  return hmac.digest("hex");
}

function SysService(
  agent: any,
  appName: string,
  serviceName: string,
  serviceKey: string
) {
  function getHash(time: number, rand: string) {
    return sha1(
      `${appName}.${serviceName}.${time}.${rand}.${serviceKey}.xdapp.com`
    );
  }

  return {
    reg(time: number, rand: string, hash: string) {
      if (agent.regSuccess)
        return {
          status: false
        };
      agent.isRegError = false;
      if (hash !== sha1(`${time}.${rand}.xdapp.com`)) {
        // 验证失败
        return {
          status: false
        };
      }

      const now = ~~(new Date().getTime() / 1000);
      if (Math.abs(now - time) > 60) {
        // 超时
        return {
          status: false
        };
      }

      return {
        status: true,
        app: appName,
        name: serviceName,
        time: now,
        rand: rand,
        version: "v1",
        hash: getHash(now, rand)
      };
    },

    regErr(msg: string) {
      agent.log(msg);
      agent.isRegError = true;
    },

    regOk(data: any, time: number, rand: string, hash: string) {
      if (agent.regSuccess) return;
      const now = new Date().getTime() / 1000;
      if (Math.abs(now - time) > 60) {
        // 超时
        agent.log(`RPC验证超时，服务名: ${appName}->${serviceName}`);
        return false;
      }

      // 不接受低于16位长度的rand
      if (rand.length < 16) {
        agent.log(
          `regOk() 回调 rand 参数太短，服务名: ${appName}->${serviceName}`
        );
        return;
      }

      if (getHash(time, rand) !== hash) {
        // 验证失败
        agent.log(`RPC验证失败，服务名: ${appName}->${serviceName}`);
        agent.socket.close();
        return;
      }

      // 注册成功
      // serviceData = data;
      agent.regSuccess = true;
      agent.serviceId = data.serviceId;

      agent.log(`RPC服务注册成功，服务名: ${appName}->${serviceName}`);

      let allService: any = {
        sys: [],
        service: [],
        other: []
      };
      agent.getNames().forEach((item: string) => {
        if (item === "#") return null;
        let pos = item.indexOf("_");
        if (pos === -1) {
          allService.other.push(item);
          return null;
        }
        let type = item.substr(0, pos);
        let func = item.substr(pos + 1);
        switch (type) {
          case "sys":
            allService.sys.push(func.replace(/_/g, ".") + "()");
            break;
          case serviceName:
            allService.service.push(func.replace(/_/g, ".") + "()");
            break;
          default:
            allService.other.push(func.replace(/_/g, ".") + "()");
            break;
        }
      });
      agent.log(`系统RPC：${allService.sys.join(", ")}`);
      agent.log(`已暴露服务RPC：${allService.service.join(", ")}`);
      if (allService.other.length > 0) {
        agent.log(`已暴露但XDApp不会调用的RPC：${allService.other.join(", ")}`);
        agent.log(
          `若需要这些方法暴露给XDApp服务使用，请加: ${serviceName} 前缀`
        );
      }
    },

    // rpc回调log输出
    log(log: any, type: any, data = null) {
      agent.log(log, type, data);
    },

    // ping
    ping() {
      return true;
    },

    // 获取服务器列表
    getFunctions(): any {
      if (!agent.regSuccess) return [];
      return agent.getNames();
    }
  }
}

export default SysService


src/xdapp.js
'use strict';

const fs = require('fs');
const tls = require('tls');
const net = require('net');
const util = require('util');
const hprose = require('hprose');
const hproseService = hprose.Service;
const Future = hprose.Future;
const sysService = require('./sysService');

const ProductionServer = {
    host: 'service-prod.xdapp.com',
    port: 8900,
};
const DevServer = {
    host: 'dev.xdapp.com',
    port: 8100
};
const GlobalServer = {
    host: 'service-gcp.xdapp.com',
    port: 8900
};

const RPC_VERSION = 1;          // RPC协议版本
const FLAG_RESULT_MODE = 2;     // 返回模式
const FLAG_FINISH = 4;          // 已完成

class XDAppServiceAgent {
    constructor(appName, serviceName, serviceKey = '') {
        if (!appName) {
            throw new Error('Required appName');
        }
        if (!serviceName) {
            throw new Error('Required serviceName');
        }
        // 注入 this  初始化 hproseService
        this.showLog = false;
        this.serviceName = serviceName;
        this.appName = appName;
        
        this.log = function(log, type) {
            console.log(log);
        };
        hproseService.call(this);

        // // 注册系统服务
        this.addInstanceMethods(sysService(this, appName, serviceName, serviceKey), 'sys');
    }

    addHttpApiProxy(url, alias = 'api', methods = ['get'], httpHeaders = []) {
        const self = this;
        if (typeof methods === 'string')methods = [methods];
        methods.forEach(function(method) {
            self.addWebFunction(function() {
                const context = self.getCurrentContext()
                const callback = arguments[arguments.length - 1]
                let uri = '', data = null, timeout = 30
                if (arguments.length >= 4) {
                    uri = arguments[0]
                    data = arguments[1]
                    timeout = arguments[2]
                }
                else if (arguments.length >= 3) {
                    uri = arguments[0]
                    data = arguments[1]
                }
                else if (arguments.length >= 2) {
                    uri = arguments[0]
                }

                let urlParse = require("url").parse(url + uri)
                const request = require('request')
                const options = {
                    uri: urlParse,
                    method: method.toUpperCase(),
                    timeout: timeout * 1000,
                    encoding: 'utf8',
                    gzip: true,
                    headers: {
                        'Host': urlParse.hostname,
                        'User-Agent': 'Chrome/49.0.2587.3',
                        'X-Xdapp-Proxy': 'True',
                        'X-Xdapp-App-Id': context.appId,
                        'X-Xdapp-Service': self.serviceName,
                        'X-Xdapp-Request-Id': context.requestId,
                        'X-Xdapp-Admin-Id': context.adminId,
                    }
                }

                if (options.method === 'POST') {
                    if (typeof data === 'object') {
                        if (options.headers['Content-Type'] && options.headers['Content-Type'].indexOf('json') !== -1) {
                            // json 方式提交在 body 里
                            options.body = JSON.stringify(data);
                        }
                        else {
                            options.headers['Content-Type'] = 'application/x-www-form-urlencoded';
                            options.form = data;
                        }
                    }
                    else {
                        options.body = '' + data;
                    }
                }
                else if (options.method === 'PUT') {
                    options.body = typeof data === 'object' ? JSON.stringify(data) : '' + data
                }

                request(options, function (error, rep, body) {
                    if (!rep)rep = {statusCode: -1}
                    if (!body)body = ''
                    callback({
                        code: rep.statusCode,
                        headers: rep.headers || {},
                        body: body,
                    })
                    self.log('Http Proxy, method: '+ method +', code: '+ rep.statusCode +', url: '+ url + uri +', headers: '+ JSON.stringify(rep.headers || {}))
                });
            }, alias + '_' + method, {async: true})
        })
    }

    // 注册一个Web浏览器可用的方法，同 hprose.addFunction 方法参数，差别是默认增加了 alias 了 serviceName 的前缀
    addWebFunction(func, alias, options) {
        if (typeof(func) !== 'function') {
            throw new Error('Argument func must be a function');
        }
        if ((options === undefined) && (typeof alias === 'object')) {
            options = alias;
            alias = null;
        }
        options = options || {};
        alias = this.serviceName + '_' + alias;

        this.addFunction(func, alias, options);
    }

    // deprecated 请使用 addWebFunction
    register(func, alias, options) {
        this.addWebFunction(func, alias, options)
    }

    // 获取当前上下文对象
    getCurrentContext() {
        return this._lastContext;
    }

    // 连接到本地测试环境
    connectToLocalDev(host = '127.0.0.1', port = 8061) {
        return this.connectTo(host, port, {
            tls: false,
            localDev: true,
            dev: true,
        });
    }

    // 连接到测试环境
    connectToDev() {
        return this.connectTo(null, null, {
            tls: true,
            localDev: false,
            dev: true,
        });
    }

    // 连接到国内生产环境
    connectToProduce() {
        return this.connectTo(null, null, {
            tls: true,
            localDev: false,
            dev: false,
        });
    }
    // 连接到海外生产环境
    connectToGlobal() {
        return this.connectTo(GlobalServer.host, GlobalServer.port, {
            tls: true,
            localDev: false,
            dev: false,
        });
    }

    // 创建一个新的连接
    connectTo(host, port, option) {
        if (this.socket)return false;

        option = Object.assign({
            tls: true,
            localDev: false,
            dev: false,
            serviceKey: null
        }, option || {});

        host = host || (option.dev ? DevServer.host : ProductionServer.host);
        port = port || (option.dev ? DevServer.port : ProductionServer.port);

        const socket = (option.tls ? tls : net).connect(port, host, {}, () => {
            this.log('Rpc client connected. ' + (socket.authorized ? 'authorized' : 'unauthorized'));
        });
        socket.on('data', (data) => {
            // CFlag/CVer/NLength/NAppId/NServiceId/NRequestId/NAdminId/CContextLength
            if (data.length < 20) {
                this.log(`接收到的数据包异常`);
                socket.end();
                return;
            }
            const flag = data.readInt8(0);
            const ver  = data.readInt8(1);
            if (ver !== RPC_VERSION)
            {
                this.log(`当前协议版本不被支持 ${ver}`);
                socket.end();
                return;
            }

            const length = data.readInt32BE(2);
            const contextLength = data.readInt8(22);
            const request = {
                flag          : flag,
                ver           : ver,
                length        : length,
                appId         : data.readInt32BE(6),
                serviceId     : data.readInt32BE(10),
                requestId     : data.readInt32BE(14),
                adminId       : data.readInt32BE(18),
                contextLength : contextLength,
                context       : contextLength > 0 ? data.slice(23, 23 + contextLength) : '',
                body          : data.slice(contextLength + 23),
            };

            const context = {
                service : this,
                client: socket,
                requestId: request.requestId,
                appId: request.appId,
                serviceId: request.serviceId,
                adminId: request.adminId,
                userdata: {}
            };
            this._lastContext = context;
            const rs = this.defaultHandle(request.body.toString(), context);

            if (Future.isFuture(rs)) {
                rs.then(function(data) {
                    _send(socket, data, request);
                });
            }
            else {
                _send(socket, data, request);
            }
        });
        socket.on('error', (error) => {
            this.log(error, 'error');
            this.socket = null;
            this.regSuccess = false;

            if (!this.isRegError)setTimeout(() => {
                // isRegError
                this.connectTo(host, port, option);
            }, 2000);
        });
        socket.on('close', () => {
            this.log('Rpc server close connection' + (this.isRegError ? '' : ', reconnect after 1 second.'));
            this.socket = null;
            this.regSuccess = false;

            if (!this.isRegError)setTimeout(() => {
                // isRegError
                this.connectTo(host, port, option);
            }, 1000);
        });

        // // 超时重置
        // socket.setTimeout(30 * 1000);
        // socket.on('timeout', () => {
        //     this.log('Rpc socket timeout.');
        //     socket.end();
        // });

        function _send(socket, body, request) {
            // CFlag/CVer/NLength/NAppId/NServiceId/NRequestId/NAdminId/CContextLength
            const packagePrefix = 6;
            const headerLength = 17;
            const length = headerLength + request.contextLength + body.length;

            const bufHeader = new Buffer.alloc(packagePrefix + headerLength);
            // packagePrefix
            const flag = request.flag | FLAG_RESULT_MODE | FLAG_FINISH;
            bufHeader.writeInt8(flag, 0);                         // flag
            bufHeader.writeInt8(RPC_VERSION, 1);                  // ver
            bufHeader.writeInt32BE(length, 2);                    // length
            // header
            bufHeader.writeInt32BE(request.appId, 6);             // appId
            bufHeader.writeInt32BE(request.serviceId, 10);        // serviceId
            bufHeader.writeInt32BE(request.requestId, 14);        // requestId
            bufHeader.writeInt32BE(request.adminId, 18);          // adminId
            bufHeader.writeInt8(request.contextLength, 22);       // contextLength

            const buffers = [bufHeader];
            if (request.contextLength > 0) {
                // 将自定义内容加入
                buffers.push(request.context);
            }
            buffers.push(body);
            const buf = Buffer.concat(buffers, bufHeader.length + request.contextLength + body.length);
            socket.write(buf);
        }

        this.socket = socket;
        return socket;
    }

    setLogHandle(handle) {
        this.log = handle;
    }
}

// 绑定 hproseService 的方法
util.inherits(XDAppServiceAgent, hproseService);

module.exports = XDAppServiceAgent;